File: backend/app/api/routes/rag.py
"""
app/api/routes/rag.py — واجهات RAG للتجربة والتشغيل المحلي

POST   /api/v1/rag/index            إضافة/تحديث مستند نصي في الفهرس (مشرف)
POST   /api/v1/rag/index/listings   فهرسة كل الإعلانات المنشورة (مشرف)
POST   /api/v1/rag/search           بحث دلالي فقط (بدون توليد)
POST   /api/v1/rag/ask              سؤال + إجابة مع المصادر (RAG كامل)
GET    /api/v1/rag/status           حالة النماذج وعدد المستندات
DELETE /api/v1/rag/documents/{id}   حذف مستند (مشرف)
"""

from __future__ import annotations

from typing import Any, Optional

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel
from sqlalchemy import text as sql
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.routes.auth import get_current_user
from app.core.database import get_db
from app.core.security import require_admin_role
from app.models.models import User
from app.services.rag_service import get_rag_service

router = APIRouter()


class IndexIn(BaseModel):
    source: str = "faq"
    source_id: Optional[str] = None
    title: Optional[str] = None
    content: str
    metadata: Optional[dict] = None


class SearchIn(BaseModel):
    query: str
    top_k: int = 5
    source: Optional[str] = None


class AskIn(BaseModel):
    question: str
    top_k: int = 5
    source: Optional[str] = None


@router.get("/status")
async def rag_status(db: AsyncSession = Depends(get_db)):
    rag = get_rag_service()
    try:
        total = (await db.execute(sql("SELECT COUNT(*) FROM rag_documents"))).scalar() or 0
        by_source = [
            {"source": r[0], "count": r[1]}
            for r in (await db.execute(sql("SELECT source, COUNT(*) FROM rag_documents GROUP BY source"))).all()
        ]
    except Exception:
        total, by_source = 0, []
    return {"ok": True, **rag.status(), "documents": total, "by_source": by_source}


@router.post("/index")
async def rag_index(
    body: IndexIn,
    _admin: User = Depends(require_admin_role("moderator")),
    db: AsyncSession = Depends(get_db),
):
    if not body.content.strip():
        raise HTTPException(400, "المحتوى فارغ")
    rag = get_rag_service()
    return await rag.upsert(
        db,
        source=body.source,
        source_id=body.source_id,
        title=body.title,
        content=body.content,
        metadata=body.metadata,
    )


@router.post("/index/listings")
async def rag_index_listings(
    _admin: User = Depends(require_admin_role("moderator")),
    db: AsyncSession = Depends(get_db),
):
    rag = get_rag_service()
    rows = (await db.execute(sql(
        """SELECT id, title, description, city, district, category, listing_type,
                  price, price_tier, condition_grade
           FROM listings WHERE status = 'published' LIMIT 500"""
    ))).all()
    indexed = 0
    for r in rows:
        content = (
            f"العنوان: {r[1] or ''}. الوصف: {r[2] or ''}. المدينة: {r[3] or ''} {r[4] or ''}. "
            f"الفئة: {r[5] or ''} - {r[6] or ''}. السعر: {r[7] or 'غير محدد'} ({r[8] or 'غير مصنف'}). "
            f"الحالة: {r[9] or 'غير محددة'}."
        )
        await rag.upsert(
            db,
            source="listing",
            source_id=str(r[0]),
            title=r[1],
            content=content,
            metadata={"city": r[3], "category": r[5], "price": float(r[7]) if r[7] else None, "tier": r[8]},
        )
        indexed += 1
    return {"indexed_listings": indexed}


@router.post("/search")
async def rag_search(
    body: SearchIn,
    _user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    if not body.query.strip():
        raise HTTPException(400, "أدخل نص البحث")
    rag = get_rag_service()
    results = await rag.search(db, body.query, body.top_k, body.source)
    return {"query": body.query, "results": results}


@router.post("/ask")
async def rag_ask(
    body: AskIn,
    _user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    if not body.question.strip():
        raise HTTPException(400, "أدخل السؤال")
    rag = get_rag_service()
    contexts = await rag.search(db, body.question, body.top_k, body.source)
    result = rag.generate(body.question, contexts)
    return {
        "question": body.question,
        **result,
        "sources": [
            {"id": c["id"], "source": c["source"], "source_id": c["source_id"], "title": c["title"], "score": c["score"]}
            for c in contexts
        ],
    }


@router.delete("/documents/{doc_id}")
async def rag_delete(
    doc_id: str,
    _admin: User = Depends(require_admin_role("admin")),
    db: AsyncSession = Depends(get_db),
):
    await db.execute(sql("DELETE FROM rag_documents WHERE id = :id"), {"id": doc_id})
    await db.commit()
    return {"deleted": doc_id}

File: server/src/routes/rag.js
import { Router } from "express";
import { db, uid, json } from "../db.js";
import { requireAuth } from "../auth.js";
import { ai } from "../ai.js";

export const router = Router();

const CHUNK = 700;

function chunkText(text) {
  const clean = String(text || "").replace(/\s+/g, " ").trim();
  const out = [];
  for (let i = 0; i < clean.length; i += CHUNK) out.push(clean.slice(i, i + CHUNK));
  return out.length ? out : [""];
}

function cosine(a, b) {
  let dot = 0, na = 0, nb = 0;
  for (let i = 0; i < Math.min(a.length, b.length); i++) {
    dot += a[i] * b[i];
    na += a[i] * a[i];
    nb += b[i] * b[i];
  }
  return dot / (Math.sqrt(na) * Math.sqrt(nb) || 1);
}

async function embedAll(texts) {
  const r = await ai.embed(texts);
  if (r?.ok && Array.isArray(r.embeddings)) return r.embeddings;
  // بديل خفيف: تمثيل حقيبة أحرف (hashing) عند غياب بايثون
  return texts.map((t) => {
    const v = new Array(256).fill(0);
    for (const ch of String(t)) v[ch.charCodeAt(0) % 256] += 1;
    const n = Math.sqrt(v.reduce((s, x) => s + x * x, 0)) || 1;
    return v.map((x) => x / n);
  });
}

router.get("/status", async (_req, res) => {
  const count = db.prepare("SELECT COUNT(*) c FROM rag_documents").get().c;
  const sources = db.prepare("SELECT source, COUNT(*) c FROM rag_documents GROUP BY source").all();
  const health = await ai.health();
  res.json({ documents: count, sources, ai: health, ready: count > 0 });
});

router.post("/index", requireAuth, async (req, res) => {
  const { source = "manual", source_id = null, title = null, content } = req.body || {};
  if (!content) return res.status(422).json({ detail: "المحتوى مطلوب" });
  const chunks = chunkText(content);
  const embeddings = await embedAll(chunks);
  const stmt = db.prepare(
    "INSERT INTO rag_documents (id, source, source_id, title, chunk_index, content, embedding) VALUES (?,?,?,?,?,?,?)",
  );
  chunks.forEach((c, i) => stmt.run(uid(), source, source_id, title, i, c, JSON.stringify(embeddings[i])));
  res.status(201).json({ indexed: chunks.length });
});

router.post("/index/listings", requireAuth, async (_req, res) => {
  const rows = db.prepare("SELECT * FROM listings WHERE status = 'published'").all();
  db.prepare("DELETE FROM rag_documents WHERE source = 'listing'").run();
  let indexed = 0;
  const stmt = db.prepare(
    "INSERT INTO rag_documents (id, source, source_id, title, chunk_index, content, embedding, metadata) VALUES (?,?,?,?,?,?,?,?)",
  );
  for (const l of rows) {
    const text = [
      l.title, l.description, l.city, l.district, l.brand, l.model,
      l.price ? `السعر ${l.price} ${l.currency}` : null,
      l.price_tier ? `تصنيف السعر ${l.price_tier}` : null,
      l.condition_grade ? `حالة السلعة ${l.condition_grade}` : null,
    ].filter(Boolean).join(" — ");
    const chunks = chunkText(text);
    const embeddings = await embedAll(chunks);
    chunks.forEach((c, i) =>
      stmt.run(uid(), "listing", l.id, l.title, i, c, JSON.stringify(embeddings[i]), JSON.stringify({ price: l.price, city: l.city })),
    );
    indexed += chunks.length;
  }
  res.json({ listings: rows.length, indexed });
});

async function retrieve(query, topK = 5, source) {
  const [qv] = await embedAll([query]);
  const rows = source
    ? db.prepare("SELECT * FROM rag_documents WHERE source = ?").all(source)
    : db.prepare("SELECT * FROM rag_documents").all();
  return rows
    .map((r) => ({
      id: r.id,
      source: r.source,
      source_id: r.source_id,
      title: r.title,
      content: r.content,
      metadata: json(r.metadata, {}),
      score: cosine(qv, json(r.embedding, [])),
    }))
    .sort((a, b) => b.score - a.score)
    .slice(0, topK);
}

router.post("/search", async (req, res) => {
  const { query, top_k = 5, source } = req.body || {};
  if (!query) return res.status(422).json({ detail: "أدخل نص البحث" });
  res.json({ query, results: await retrieve(query, Number(top_k), source) });
});

router.post("/ask", async (req, res) => {
  const { question, top_k = 5, source } = req.body || {};
  if (!question) return res.status(422).json({ detail: "أدخل سؤالك" });
  const results = await retrieve(question, Number(top_k), source);
  const gen = await ai.generate(question, results.map((r) => r.content));
  const answer =
    gen?.ok && gen.answer
      ? gen.answer
      : results.length
        ? `استناداً إلى البيانات المتاحة:\n\n${results.slice(0, 3).map((r, i) => `${i + 1}) ${r.content}`).join("\n")}`
        : "لا توجد بيانات مفهرسة كافية للإجابة. جرّب فهرسة الإعلانات أولاً.";
  res.json({ question, answer, sources: results });
});

router.delete("/documents/:id", requireAuth, (req, res) => {
  db.prepare("DELETE FROM rag_documents WHERE id = ? OR source_id = ?").run(req.params.id, req.params.id);
  res.json({ deleted: req.params.id });
});
